### PR TITLE
Allow analyzing manifests and particular versions

### LINKED ISSUE #70: Analyze manifest
I think it would be nice to analyze all the packages in a Manifest.toml. Parsing the Manifest is easy & getting the list of packages is easy enough, but the tricky part is if you want to actually analyze the versions of the packages that are encoded in that manifest.

Why is it tricky? Well, currently, we always do a shallow clone of the repo and analyze it's default branch. But that may be very different than the version used in a manifest. We can get git tree sha out of the manifest, which is a tree hash encoding the directory tree of the code being used, and we can use that + a full clone, or with the github API to get a tarball of the directory tree for the code, which we can analyze.

BUT subdir packages complicate this: many of the things we check for only exist at the top-level of most repos, e.g. CI workflows and docs. So the package code tarball doesn't include that stuff. So we actually do depend on the state of the repo outside the tree hash. But that state isn't actually determined by the tree hash of the code alone: if a particular subdirectory tree (with the code) never changes as CI workflows etc change outside of it, we can't map exactly from tree-hash of code to tree-hash of full repo (or commit hash of the full repo). Probably we can find some commits that include this tree hash, but we don't know which to use-- and in fact, the registry does not include information about this.

So, what to do. I think probably one way to move forward would be to give up on learning top-level information on subdirs when we are trying to analyze a particular tree hash.

### PR DESCRIPTION
**Pull request by ericphanson**
Closes #70 

Unsurprisingly, correctly handling the cases of non-release packages added by `Pkg.add`, packages added by `Pkg.dev`, and arbitrary release versions possibly spread over multiple registries has proven somewhat complicated. In order to have some hope to tame the complexity, I have refactored much of the code. I am hoping it can be somewhat clear in the end, with some organization and comments.

This does have a lot of nice features though:
* `analyze_manifest` allows analyzing all the code specified in a manifest
* `tree_hash`'s are checked post-download, when available (i.e. not a dev'd package). Corrupted but non-error'd download should count as not-reachable, instead of getting garbage results.
* code from installed packages (`.julia/packages/name/slug`) is analyzed instead of being re-downloaded (and is depot-agnostic)
    * in particular, if you have an `instantiate`'d manifest, no code needs to be downloaded to analyze it. It's fast!
* `analyze(name)` now defaults to the latest release version, instead of the latest trunk. In particular, downloads can be skipped if this version is installed.
    * `analyze` allows choosing a version, including `:stable` for latest release and `:dev` for unreleased trunk branch
* handles multiple registries. Packages can be looked up in any reachable registry, and even correctly supports the same package being registered in multiple registries, possibly w different versions in each
* code paths are somewhat streamlined. We always flow from "input" (name, url, directory, module, etc) to `PkgSource` (`Release`, `Dev`, `Added`, or `Trunk` representing the valid ways to access a package), with `obtain_code` to obtain the code for each possible source, to `analyze_code` which is agnostic to how the code was obtained.
    * release/dev/added are ways to get a package in a manifest. `Trunk` lets you pass just a URL (and subdir) to get the latest version. This sounds like what `dev` does with a URL, but the implementation is different; Pkg.dev will clone it then update the manifest with a local path. So `Dev` objects track local paths, while `Trunk` objects track remote URLs.

TODO:
- [x] try to handle non-release versions in Manifest.tomls
- [x] handle dev with remote URL & subdir ?
- [x] cleanup
- [x] docs update
- [ ] stretch: try to resolve stdlibs relative to version-in-manifest when analyzing manifest, instead of current julia version
- [x] lots of tests -- many new features, some fairly complicated situations, needs a lot of tests. Current tests will fail due to refactoring.
    - [x] `Base.show` for `Release`, `Added`, `Trunk`, `Dev`
    - [x] analyze `Added` from URL
    - [x] analyze `Added` with path
    - [x] analyze `Dev` with path
    - [x] analyze `Trunk`
    - [x] analyze `Release`
    - [x] zero-arg `analyze`
    - [x] `find_packages_in_manifest` with non-release packages
    - [x] ...maybe multi-registry stuff. I think this can be tested without tooo much effort by faking a 2nd RegistryInstance by copying and modifying the one associated to General.

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -1,207 +1,385 @@
 using Test, UUIDs
 using PackageAnalyzer
-using PackageAnalyzer: parse_project
+using PackageAnalyzer: parse_project, Release, Added, Dev, Trunk, PkgSource
 using JLLWrappers
-using GitHub
+using GitHub: GitHub
```
Comment (giordano) [inline test/runtests.jl:5]: What's the point of this?

File: test/runtests.jl [hunk 2]
```diff
@@ -1,207 +1,385 @@
...
+        @test all(p -> p isa Release, few)
+        results = analyze_packages(few; auth)
+        @test length(results) == length(few)
+        # Version number saved out
+        @test all(results[i].version == few[i].version for i in 1:length(few))
+    end
+
+    @testset "`analyze`" begin
+        # `dev` analysis:
+        for pkg in (analyze(PackageAnalyzer; auth), analyze("https://github.com/giordano/PackageAnalyzer.jl"; auth), analyze(joinpath(@__DIR__, ".."); auth))
```
Comment (giordano) [inline test/runtests.jl:77]: I guess we can update the URL now :sweat_smile: 
```suggestion
        for pkg in (analyze(PackageAnalyzer; auth), analyze("https://github.com/JuliaEcosystem/PackageAnalyzer.jl"; auth), analyze(joinpath(@__DIR__, ".."); auth))
```

**Review by giordano — APPROVED** [review]
Sorry for the late review, I started it too many times, never finished and always mostly restart from scratch! :disappointed:

I went with a quick review now.  It looks overall great, tests are passing, and I like the new `@debug`s!  Needs a rebase now though

File: test/runtests.jl (hunk 1 above)
Comment (ericphanson) [reply test/runtests.jl:5]:
```
WARNING: using GitHub.Release in module Main conflicts with an existing identifier.
```

Comment (giordano) [reply test/runtests.jl:5]: I guess my question is: how's this different from `import GitHub`?

Comment (ericphanson) [reply test/runtests.jl:5]: Oh, it’s not. YAS style guide requires this form so I am used to writing it, but it’s the same

Comment (giordano) [reply test/runtests.jl:5]: Uhm, I find it a bit bizarre to be honest, but it's ok I guess.

Comment (ericphanson) [reply test/runtests.jl:5]: We can change it, I don’t care either way:
```suggestion
import GitHub
```

**Pull request merged**